server/rest/src/main/java/org/infinispan/rest/resources/CounterResource.java

```diff
@@ -30,84 +50,299 @@
+
+            // Counter ops
             .invocation().methods(GET).path("/v2/counters/{counterName}").handleWith(this::getCounter)
-            .invocation().methods(DELETE).path("/v2/counters/{counterName}").handleWith(this::resetCounter)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("reset").handleWith(this::resetCounter)
+
+            // Weak counter ops
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("increment").handleWith(this::incrementWeak)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("decrement").handleWith(this::decrementWeak)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("add").handleWith(this::addValueWeak)
+
+            // Strong counter ops
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("incrementAndGet").handleWith(this::incrementStrong)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("decrementAndGet").handleWith(this::decrementStrong)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("addAndGet").handleWith(this::addValueStrong)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("compareAndSet").handleWith(this::compareSet)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("compareAndSwap").handleWith(this::compareSwap)
             .create();
    }
 
+   private CompletionStage<RestResponse> createCounter(RestRequest request) throws RestResponseException {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+      String counterName = request.variables().get("counterName");
+
+      String contents = request.contents().asString();
+
+      if (contents == null || contents.length() == 0) {
+         responseBuilder.status(HttpResponseStatus.BAD_REQUEST);
+         responseBuilder.entity("Configuration not provided");
+         return completedFuture(responseBuilder.build());
+      }
+      try {
+         CounterConfiguration configuration = createCounterConfiguration(contents);
+         if (configuration == null) {
+            responseBuilder.status(HttpResponseStatus.BAD_REQUEST);
+            responseBuilder.entity("Invalid configuration");
+            return completedFuture(responseBuilder.build());
+         }
+         return counterManager.defineCounterAsync(counterName, configuration).thenApply(r -> new NettyRestResponse.Builder().build());
+      } catch (IOException e) {
+         throw new RestResponseException(e.getCause());
+      }
+   }
+
+   private CounterConfiguration createCounterConfiguration(String json) throws IOException {
+      JsonNode jsonNode = invocationHelper.getMapper().readValue(json, JsonNode.class);
+      boolean strongCounter = jsonNode.has("strong-counter");
+      boolean weakCounter = jsonNode.has("weak-counter");
```

Comment:
Having to check for this value in the JSON object makes me think it would be better to use a URL scheme like `v2/counters/strong/{counterName}`. It reduces the verbosity of the JSON payload and provides a clear separation of weak/strong counters.

Reply:
The issue is, the configuration schema itself contains ```weak-counter``` and ```strong-counter```.
Here, we are doing the same strategy as caches: the user provides a name and an config, and if it turns out to be a dist, repl or local cache is guided bu this config

Reply:
I see, I didn't realise this was already established for caches.